File: legacy_python/app/demos/race_condition.py
"""
Demo 1 — Race Condition

What it shows:
  Two threads each try to increment the same counter many times. Without a
  lock, some increments are lost because `counter += 1` is not atomic in
  Python (it is read-modify-write at the bytecode level under free-threaded
  builds, and the GIL release granularity still allows interleavings here
  via explicit yields).

Why it matters in RTOS:
  Race conditions corrupt shared state silently. Real-time tasks share
  buffers, hardware registers, and counters; lost updates can crash a
  control loop or, worse, look fine in testing and fail in production.
"""

import threading
import time

from .base import Sender, make_bridge


# We force the race by using a small sleep inside the increment so the
# scheduler reliably switches threads mid-update. A pure `+=` works on
# free-threaded Python but is hard to reproduce on standard CPython.
def _do_increments(counter_ref: list, lock, n: int, use_lock: bool) -> None:
    for _ in range(n):
        if use_lock:
            with lock:
                # Read, modify, write. Two threads inside this block at the
                # same time would both read the same value and one update
                # would be lost.
                v = counter_ref[0]
                counter_ref[0] = v + 1
        else:
            v = counter_ref[0]
            # Yield the CPU on purpose so the other thread can squeeze in
            # between the read and the write — this is what a real race
            # looks like, just sped up so you can see it in 1 second.
            time.sleep(0)
            counter_ref[0] = v + 1


def _experiment(n_per_thread: int, use_lock: bool) -> dict:
    counter_ref = [0]
    lock = threading.Lock()
    threads = [
        threading.Thread(target=_do_increments,
                         args=(counter_ref, lock, n_per_thread, use_lock))
        for _ in range(2)
    ]
    t0 = time.perf_counter()
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    elapsed_ms = (time.perf_counter() - t0) * 1000.0
    expected = n_per_thread * 2
    actual = counter_ref[0]
    return {
        "expected": expected,
        "actual": actual,
        "lost": expected - actual,
        "elapsed_ms": round(elapsed_ms, 1),
    }


async def run(send: Sender, params: dict) -> None:
    """
    Run two passes back-to-back: first without a lock, then with one.
    The UI shows both side-by-side so the difference is obvious.
    """
    n = int(params.get("iterations", 50_000))
    bridge = make_bridge(send)

    # Pass 1 — no lock
    await send({"type": "phase", "label": "Without mutex (broken)"})
    unsafe = await _run_in_thread(_experiment, n, False)
    await send({"type": "result", "variant": "unsafe", **unsafe})

    # Pass 2 — with lock
    await send({"type": "phase", "label": "With mutex (correct)"})
    safe = await _run_in_thread(_experiment, n, True)
    await send({"type": "result", "variant": "safe", **safe})

    await send({"type": "done"})


async def _run_in_thread(fn, *args):
    import asyncio
    return await asyncio.to_thread(fn, *args)

File: docs/demos/race_condition.js
/* Demo 1 — Race condition.
 *
 * Two real Web Workers share a SharedArrayBuffer. We run two passes:
 *   (a) without any synchronization — the count is wrong.
 *   (b) with an Atomics-based mutex   — the count is exact.
 *
 * Atomics.wait / Atomics.notify are the JS equivalent of Linux futex(2).
 */

const ITERATIONS = 60_000;

function spawnPair(mode, iterations) {
  return new Promise((resolve) => {
    const sab = new SharedArrayBuffer(3 * 4);   // 3 × Int32
    const view = new Int32Array(sab);
    view[0] = 0; view[1] = 0; view[2] = 0;

    const w1 = new Worker('demos/worker_counter.js');
    const w2 = new Worker('demos/worker_counter.js');
    let returned = 0;
    let totalElapsed = 0;

    function onDone(e) {
      returned += 1;
      totalElapsed = Math.max(totalElapsed, e.data.elapsed);
      if (returned === 2) {
        w1.terminate(); w2.terminate();
        resolve({
          actual: view[0],
          expected: iterations * 2,
          elapsedMs: totalElapsed,
        });
      }
    }
    w1.onmessage = onDone;
    w2.onmessage = onDone;

    w1.postMessage({ sab, iterations, mode, threadId: 1 });
    w2.postMessage({ sab, iterations, mode, threadId: 2 });
  });
}

export async function run(params, send) {
  const iterations = params?.iterations ?? ITERATIONS;

  await send({ type: 'phase', label: 'Without mutex (broken)' });
  const unsafe = await spawnPair('unsafe', iterations);
  await send({
    type: 'result', variant: 'unsafe',
    expected: unsafe.expected,
    actual: unsafe.actual,
    lost: unsafe.expected - unsafe.actual,
    elapsed_ms: Math.round(unsafe.elapsedMs * 10) / 10,
  });

  await send({ type: 'phase', label: 'With mutex (correct)' });
  const safe = await spawnPair('mutex', iterations);
  await send({
    type: 'result', variant: 'safe',
    expected: safe.expected,
    actual: safe.actual,
    lost: safe.expected - safe.actual,
    elapsed_ms: Math.round(safe.elapsedMs * 10) / 10,
  });

  await send({ type: 'done' });
}
